MaxText/decode_checkpoint.py

```diff
@@ -0,0 +1,131 @@
+
+def read_training_checkpoint(config, checkpoint_manager):
+  is_training = True
+  return _read_checkpoint(config, checkpoint_manager, is_training)
+
+def _save_decode_checkpoint(config, state, checkpoint_manager):
+  """Generate checkpoint for decode from the training_state."""
+  with jax.spmd_mode('allow_all'):
+    decode_state = max_utils.init_decode_state(state.apply_fn, state.params)
+  if checkpoint_manager is not None:
+    if checkpoint_manager.save(0, decode_state):
+      max_logging.log(f"saved an decode checkpoint at {config.checkpoint_dir}")
+    # Upon preemption, exit when and only when all ongoing saves are complete.
+    if checkpoint_manager.reached_preemption(0):
+      checkpoint_manager.wait_until_finished()
+      sys.exit()
+
+def generate_decode_checkpoint(config):
+  """
+  Generate an decode checkpoint from a given training checkpoint.
+  - Training checkpoint is saved at config.load_parameters_path.
+  - Inference checkpoint will be saved at config.checkpoint_dir.
+  """
+  assert config.checkpoint_dir, "checkpoint_dir not configured"
+  # Remove any old checkpoint
+  path = epath.Path(config.checkpoint_dir)
+  if path.exists():
+    if jax.process_index() == 0:
+      path.rmtree()
+
+  assert config.load_parameters_path, "load_parameters_path not configured"
+  assert epath.Path(config.load_parameters_path).exists(), "no checkpoint at load_parameters_path"
+
+  # Create a checkpoint manager to save decode checkpoint at config.checkpoint_dir
+  checkpoint_manager = checkpointing.create_orbax_checkpoint_manager(
+      config.checkpoint_dir,
+      config.enable_checkpointing,
+      config.async_checkpointing,
+      config.save_period,
+  )
+  # Read training state from config.load_paramaters_path
+  max_logging.log(f"Read training checkpoint from: {config.load_parameters_path}")
+  training_state, _ = read_training_checkpoint(config, checkpoint_manager)
+  assert training_state.opt_state != {}, "missing opt_state in training checkpoint"
+
+  # Save decode state to config.checkpoint_dir at step 0
+  max_logging.log(f"Save decode checkpoint at: {config.checkpoint_dir}")
+  _save_decode_checkpoint(config, training_state, checkpoint_manager)
+
+  # Read  newly saved decode state from config.checkpoint_dir at step 0
+  max_logging.log(f"Read decode checkpoint from: {config.checkpoint_dir}")
+  decode_state, _ = read_decode_checkpoint(config, checkpoint_manager)
+  assert decode_state.opt_state == {}, "non null opt_state in decode checkpoint"
+  num_training_params = max_utils.calculate_num_params_from_pytree(training_state.params)
+  num_decode_params = max_utils.calculate_num_params_from_pytree(decode_state.params)
+  assert num_training_params == num_decode_params, "count mismatch in params in checkpoints"
+
+  max_logging.log(f"Successfully generated decode checkpoint at: {config.checkpoint_dir}0/default")
+  return True
+
+
+def main(argv: Sequence[str]) -> None:
```

Comment:
I believe this function's only job is to generate_decode_checkpoint's in which case I think we should call it generate_decode_checkpoint.py.